core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -1211,37 +1033,30 @@ object GroupMetadataManager {
   def offsetCommitValue(offsetAndMetadata: OffsetAndMetadata,
                         apiVersion: ApiVersion): Array[Byte] = {
     // generate commit value according to schema version
-    val (version, value) = {
+    val (version, value) =
```

Comment:
I wonder if we could simplify the whole logic by marking the `ExpireTimestamp` and `LeaderEpoch` as optional in the schema. That would allow to always set them here. The auto-generated protocol would then use it only for the supported versions. The logic to choose the version would remain identical:

```
val version = 
  if (apiVersion < KAFKA_2_1_IV0 || offsetAndMetadata.expireTimestamp.nonEmpty) 1.toShort
  else if (apiVersion < KAFKA_2_1_IV1) 2.toShort
  else 3.toShort
```

Reply:
great idea!